datumaro/cli/commands/download.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) 2021 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import argparse
+import logging as log
+import os
+import os.path as osp
+
+from datumaro.components.extractor_tfds import (
+    AVAILABLE_TFDS_DATASETS, TFDS_EXTRACTOR_AVAILABLE, make_tfds_extractor,
+)
+from datumaro.components.project import Environment
+from datumaro.util.os_util import make_file_name
+
+from ..util import MultilineFormatter
+from ..util.errors import CliException
+from ..util.project import generate_next_file_name
+
+
+def build_parser(parser_ctor=argparse.ArgumentParser):
+    builtin_writers = sorted(Environment().converters)
+    if TFDS_EXTRACTOR_AVAILABLE:
+        available_datasets = ", ".join(
+            f'tfds:{name}' for name in AVAILABLE_TFDS_DATASETS)
+    else:
+        available_datasets = "N/A (TensorFlow and/or TensorFlow Datasets " \
+            "are not installed)"
+
+    parser = parser_ctor(
+        help="Download a publicly available dataset",
+        description="""
+        Downloads a publicly available dataset and saves it in a given format.|n
+        |n
+        Currently, the only source of datasets is the TensorFlow Datasets
+        library. Therefore, to use this command you must install TensorFlow &
+        TFDS, which you can do as follows:|n
+        |n
+        |s|spip install datumaro[tf,tfds]|n
+        |n
+        Supported datasets: {}|n
+        |n
+        For information about the datasets, see the TFDS Catalog:
+        <https://www.tensorflow.org/datasets/catalog/overview>.|n
+        |n
+        Supported output formats: {}|n
+        |n
+        Examples:|n
+        - Download the MNIST dataset, saving it in the ImageNet text format:|n
+        |s|s%(prog)s -i tfds:mnist -f imagenet_txt -- --save-images|n
+        |n
+        - Download the VOC 2012 dataset, saving only the annotations in the COCO
+          format into a specific directory:|n
+        |s|s%(prog)s -i tfds:voc/2012 -f coco -o path/I/like/
+        """.format(available_datasets, ', '.join(builtin_writers)),
+        formatter_class=MultilineFormatter)
+
+    parser.add_argument('-i', '--dataset-id', required=True,
+        help="Which dataset to download")
+    parser.add_argument('-f', '--output-format', required=True,
+        help="Output format")
```

Comment:
What is the point of having this option is the command? From my point, it should be pretty straightforward that we download a dataset and save it in the same format. Why do we need this separation? I can see it as a shortcut for calling `download` + `convert` (which should not improve performance) and as a way to support loading from custom storages (which is questionable).

Reply:
The original format of the dataset is lost once it goes through TFDS. By the time we get the data, it's transformed into a generic form, so it makes as much sense to save it in the original format as any other format.

In addition, TFDS contains a wide variety of datasets, and we might want to implement downloading datasets for which Datumaro itself has no support (or no write support). The only way we'd be able to save those is with conversion.

Reply:
> The original format of the dataset is lost once it goes through TFDS. By the time we get the data, it's transformed into a generic form, so it makes as much sense to save it in the original format as any other format.

I'd suggest to go from the task and not from the implementation. If I write
```bash
datum download mnist
```
I feel it's pretty reasonable for this command to work without specification of the output format (at least, if it is supported in DM). We know what dataset we're downloading, and TFDS datasets probably can be mapped to DM formats without ambiguity. Can we make this parameter optional by maintaining a table of default formats for tfds ids? Output conversion on the fly can be useful in some cases, but I don't understand, why it must be required.

Reply:
> Can we make this parameter optional by maintaining a table of default formats for tfds ids?

I can do that, but keep in mind that conversion will still happen behind the scenes. It'll just be conversion to that default format.

> datum download mnist

Incidentally, one of the reasons why the command does _not_ look like that (you have to specify `tfds:mnist`) is that the output is not the original MNIST dataset, but MNIST as it is represented in TFDS. There may be some loss of information incurred by TFDS's own parsing/conversion code (for example, the original order of items is lost, because TFDS shuffles the items before saving them).

Reply:
> I can do that, but keep in mind that conversion will still happen behind the scenes. It'll just be conversion to that default format.

I understand, but in this variant the user is less distracted with an extra mandatory parameter. 

> output is not the original MNIST dataset, but MNIST as it is represented in TFDS

Sure, but it was not the point I wanted to highlight - all the same applies to `datum download tfds:mnist`. The key here is that the (default) output format should be predefined. One can argue if it should be tfrecord or just MNIST, I believe it should be MNIST, because (1) the TFDS dataset will still be available in the cache and (2) it allows to just start using the dataset immediately.

> for example, the original order of items is lost, because TFDS shuffles the items before saving them

Yes, I noticed that, is there an option to disable this?

Reply:
Okay, I will add default output formats.

> Yes, I noticed that, is there an option to disable this?

Not on the client side. It's controlled by the dataset implementation (see <https://www.tensorflow.org/datasets/add_dataset#maintain_dataset_order>), and last time I checked, none of the built-in datasets disabled it.

Reply:
Implemented.